### PR TITLE
[Discover] Data Grid Pagination Options

### LINKED ISSUE #5569: [Features] Configurable rows per page in the discover view
**Is your feature request related to a problem? Please describe.**

The new discover view splits search results into pages instead of automatically loading new pages as you scroll downwards. Looking through large search results were much easier when not having to navigate through pages and having 100 results per page isn't very much when dealing with large log files. It also restricts quick searching back with Ctrl+F in the browser if things are no longer on the same page.

**Describe the solution you'd like**

I'd like to be able to be able to either switch back to the old behaviour, or at least be able to configure the number of results per page as an advanced setting to increase the results shown at the same time.

**Additional context**

Inspecting the Network page of the browser DevTools it seems that all results are fetched in one request, so I'm not even sure what purpose the pages serve if everything is returned at the same time anyway?

### LINKED ISSUE #5601: Discover Page : Pagination vs scroll ( older way)
**Problem Statement:**
The older version of Open search before new discover ( with Angular ) used to have scroll of 500 documents in the same page. Some of our users use that feature extensively for triaging issues. With pagination, the context is lost .

**Solution proposed**
1. Make this configuration based if scroll vs pagination.
2. Increase the pagination size to 200 or 300 than limiting to 100.

I understand it might be related to new UI working across.

**Alternatives**
No alternatives are considered for now. We have asked users to refine search better to have 100 records.

### PR DESCRIPTION
**Pull request by kishor82**
### Description

Implemented pagination options based on the Discover UI setting configuration value. #5601 , #5569

## Screenshot
1. Number of rows is set to 1000
Screenshot 2023-12-14 at 4 45 44 AM

2. Pagination options are updated based on settings
Screenshot 2023-12-14 at 4 46 57 AM

closes #5601
closes #5569

### Check List

- [X] All tests pass
  - [X] `yarn test:jest`
  - [X] `yarn test:jest_integration`
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
- [X] Update [CHANGELOG.md](./../CHANGELOG.md)
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -945,4 +946,4 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 ### 🔩 Tests
 
-- Update caniuse to fix failed integration tests ([#2322](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2322))
+- Update caniuse to fix failed integration tests ([#2322](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2322))
```
Comment (kishor82) [inline CHANGELOG.md:949]: Don't know what changed here!

**Review by ananzh — APPROVED** [review]
nice! thank you

File: src/plugins/discover/public/application/components/utils/use_pagination.ts [hunk 2]
```diff
@@ -4,13 +4,25 @@
...
 import { useState, useMemo, useCallback } from 'react';
+import { DiscoverServices } from '../../../build_services';
+import { SAMPLE_SIZE_SETTING } from '../../../../common';
+import { generatePageSizeOptions } from './page_size_options';
+export interface Props {
+  services: DiscoverServices;
+  rowCount: number;
+}
+
+export const usePagination = ({ rowCount, services }: Props) => {
```
Comment (ruanyl) [inline src/plugins/discover/public/application/components/utils/use_pagination.ts:15]: I'd avoid passing the entire `services` object, but passing `pageSizeLimit` here which is `sampleSize` in this case. Does it make sense?

```suggestion
export const usePagination = (rowCount: number, pageSizeLimit: number) => {
```

File: src/plugins/discover/public/application/components/utils/page_size_options.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * @returns {number[]} - An array of available page size options.
+ */
+
+export const generatePageSizeOptions = (sampleSize: number): number[] => {
+  if (sampleSize && sampleSize > 0) {
+    const maxSize = 500;
+    const pageSizeFromSetting = [...Array(Math.floor(sampleSize / maxSize)).keys()].map(
+      (i) => (i + 1) * maxSize
+    );
+    return uniq([...defaultPageOptions, ...pageSizeFromSetting]).sort((a, b) => a - b);
```
Comment (ruanyl) [inline src/plugins/discover/public/application/components/utils/page_size_options.ts:22]: We can replace `uniq` with `Array.from(new Set())`, we're getting rid of `lodash` as I know.

File: src/plugins/discover/public/application/components/utils/page_size_options.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * The array includes default values (25, 50, 100) and additional options derived from the `sampleSize` setting.
+ * Ensures uniqueness and sorts the array in ascending order, representing available page size options for pagination.
+ * @param {number} sampleSize - The sample size used to determine additional pagination options.
+ * @returns {number[]} - An array of available page size options.
+ */
+
+export const generatePageSizeOptions = (sampleSize: number): number[] => {
+  if (sampleSize && sampleSize > 0) {
+    const maxSize = 500;
+    const pageSizeFromSetting = [...Array(Math.floor(sampleSize / maxSize)).keys()].map(
```
Comment (ruanyl) [inline src/plugins/discover/public/application/components/utils/page_size_options.ts:19]: It seems the page size options is `[25, 50, 100]` when `sampleSize` is `400`, is it expected?

File: src/plugins/discover/public/application/components/utils/page_size_options.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * Generates an array of pagination options based on the provided `sampleSize`.
+ * The array includes default values (25, 50, 100) and additional options derived from the `sampleSize` setting.
+ * Ensures uniqueness and sorts the array in ascending order, representing available page size options for pagination.
+ * @param {number} sampleSize - The sample size used to determine additional pagination options.
+ * @returns {number[]} - An array of available page size options.
+ */
+
+export const generatePageSizeOptions = (sampleSize: number): number[] => {
+  if (sampleSize && sampleSize > 0) {
+    const maxSize = 500;
```
Comment (ruanyl) [inline src/plugins/discover/public/application/components/utils/page_size_options.ts:18]: Will `stepSize` sound more appropriate?

**Comment by ananzh** [comment]
@kishor82 I think @ruanyl make great suggestions here. Could you solve the comment and update the CHANGELOG?

File: src/plugins/discover/public/application/components/utils/use_pagination.ts (hunk 2 above)
Comment (kishor82) [reply src/plugins/discover/public/application/components/utils/use_pagination.ts:15]: yes, thanks for the suggestions.

File: src/plugins/discover/public/application/components/utils/page_size_options.ts (hunk 4 above)
Comment (kishor82) [reply src/plugins/discover/public/application/components/utils/page_size_options.ts:19]: No, the idea was to provide a single page for a large dataset as per the issue. I think replacing 'floor' with 'ceil' works as expected. if you have any suggestions let me know.

File: src/plugins/discover/public/application/components/utils/page_size_options.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Generates an array of pagination options based on the provided `stepSize`.
+ * The array includes default values (25, 50, 100) and additional options derived from the `stepSize` setting.
+ * Ensures uniqueness and sorts the array in ascending order, representing available page size options for pagination.
+ * @param {number} stepSize - The sample size used to determine additional pagination options.
+ * @returns {number[]} - An array of available page size options.
+ */
+
+export const generatePageSizeOptions = (stepSize: number): number[] => {
```
Comment (ruanyl) [inline src/plugins/discover/public/application/components/utils/page_size_options.ts:14]: I think it should be called `pageSizeLimit` here: 
```suggestion
export const generatePageSizeOptions = (pageSizeLimit: number): number[] => {
```

And  `const stepSize = 500;` on line 18 :)

Comment (kishor82) [reply src/plugins/discover/public/application/components/utils/page_size_options.ts]: Ohh, I changed the wrong one.

**Review by ruanyl — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]

**Comment by kishor82** [comment]
Working on functional testing; those are failing UI setting values.

**Comment by ananzh** [comment]
@kishor82 thank you. Feel free to reach out if you need any help for functional tst.

**Comment by kishor82** [comment]
> @kishor82 thank you. Feel free to reach out if you need any help for functional tst.

@ananzh I resolved the functional test issue, and initially, all tests were passing. However, it is currently failing again after the merge. will reach out if needed.

**Review by ashwin-pc — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Pull request merged**